prow/crier/controller.go

```diff
@@ -0,0 +1,157 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// package crier reports finished prowjob status to git providers
+package crier
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+)
+
+// Controller struct defines how a controller should encapsulate
+// logging, client connectivity, informing (list and watching)
+// queueing, and handling of resource changes
+type Controller struct {
+	clientset kubernetes.Interface
+	queue     workqueue.RateLimitingInterface
+	informer  cache.SharedIndexInformer
+	handler   Handler
+}
+
+func NewController(clientset kubernetes.Interface, queue workqueue.RateLimitingInterface, informer cache.SharedIndexInformer, handler Handler) *Controller {
+	return &Controller{
+		clientset: clientset,
+		queue:     queue,
+		informer:  informer,
+		handler:   handler,
+	}
+}
+
+// Run is the main path of execution for the controller loop
+func (c *Controller) Run(stopCh <-chan struct{}) {
+	// handle a panic with logging and exiting
+	defer utilruntime.HandleCrash()
+	// ignore new items in the queue but when all goroutines
+	// have completed existing items then shutdown
+	defer c.queue.ShutDown()
+
+	logrus.Info("Controller.Run: initiating")
+
+	// run the informer to start listing and watching resources
+	go c.informer.Run(stopCh)
+
+	// do the initial synchronization (one time) to populate resources
+	if !cache.WaitForCacheSync(stopCh, c.HasSynced) {
+		utilruntime.HandleError(fmt.Errorf("Error syncing cache"))
+		return
+	}
+	logrus.Info("Controller.Run: cache sync complete")
+
+	// run the runWorker method every second with a stop channel
+	wait.Until(c.runWorker, time.Second, stopCh)
+}
+
+// HasSynced allows us to satisfy the Controller interface
+// by wiring up the informer's HasSynced method to it
+func (c *Controller) HasSynced() bool {
+	return c.informer.HasSynced()
+}
+
+// runWorker executes the loop to process new items added to the queue
+func (c *Controller) runWorker() {
+	logrus.Info("Controller.runWorker: starting")
+
+	// invoke processNextItem to fetch and consume the next change
+	// to a watched or listed resource
+	for c.processNextItem() {
+		logrus.Info("Controller.runWorker: processing next item")
+	}
+
+	logrus.Info("Controller.runWorker: completed")
+}
+
+// processNextItem retrieves each queued item and takes the
+// necessary handler action based off of if the item was
+// created or deleted
+func (c *Controller) processNextItem() bool {
```

Comment:
This seems a lot more complex than what @fejta and I have seen... where is it from? Why?

Reply:
:-) I basically copied the controller code from https://github.com/trstringer/k8s-controller-custom-resource/blob/master/controller.go#L72, I'm sure I can get rid of something

Reply:
https://github.com/kubernetes/sample-controller/blob/25b3dd894771d72289e316f63262499e7c89323a/controller.go#L188 ?